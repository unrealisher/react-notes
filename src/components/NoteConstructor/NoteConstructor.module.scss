.background {
  position: fixed;
  width: 100vw;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: rgba(218, 220, 224, 0.5);
  z-index: 10;
}

.popup {
  position: relative;
  padding: 24px;
  height: calc(100vh - 48px);
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background-color: rgb(223, 165, 165);
  border: 1px solid rgba(218, 220, 224, 0.5);
  border-radius: 15px;
}

@media (min-width: 548px) {
  .popup {
    max-width: 500px;
    height: auto;
  }
}

.exit {
  cursor: pointer;
  position: absolute;
  background-color: transparent;
  width: 25px;
  top: 10px;
  left: calc(100% - 40px);
  height: 25px;
  opacity: 0.4;

  &::before,
  &::after {
    content: "";
    position: absolute;
    top: calc(50% - 1px);
    left: 0;
    width: 24px;
    height: 2px;
    background-color: black;
  }

  &::before {
    transform: rotate(45deg);
  }

  &::after {
    transform: rotate(-45deg);
  }

  &:hover {
    opacity: 1;
  }
}

.title_wrapper {
  display: flex;
  width: 100%;
  margin: 12px 0;
}

.title {
  margin-right: 5px;
}

.title_input {
  flex-grow: 1;
  padding: 2px 10px;
  border-radius: 15px;
}

.button {
  margin-top: 12px;
  padding: 5px 10px;
  border-radius: 15px;
  border: 1px solid rgba(218, 220, 224, 0.5);
  background-color: white;

  &:hover,
  &:focus {
    background-color: #fc0;
  }
}

.wrapper {
  margin-bottom: 12px;
}
