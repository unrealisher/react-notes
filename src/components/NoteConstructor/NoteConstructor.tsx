import React, { useState, useRef, useEffect } from "react";

import TypeList from "./../TypeList/TypeList";

import styles from "./NoteConstructor.module.scss";
import TextConstructor from "../TextConstructor/TextConstructor";
import TagsList from "../TagsList/TagsList";
import ListConstructor from "../ListConstructor/ListConstructor";
import ImageConstructor from "../ImageConstructor/ImageConstructor";
import ConstructorColors from "../ConstructorColors/ConstructorColors";
import INote from "../../interfaces/INote";
import ConstructorReminder from "../ConstructorReminder/ConstructorReminder";
import ConstructorAttachments from "../ConstructorAttachments/ConstructorAttachments";
import { connect } from "react-redux";
import { AnyAction } from "redux";
import IState from "../../interfaces/IState";
import { patchNote } from "../../store/actions/patchNote";
import { ThunkDispatch } from "redux-thunk";
import { addNote } from "../../store/actions/addNote";

interface IDispatchToProps {
  onAddItem: Function;
  onPatchItem: Function;
}

interface IProps extends IDispatchToProps {
  note: INote;
  setPopup: Function;
  setPatchItem: Function;
}

export const NoteConstructor = (props: IProps): JSX.Element => {
  const { setPopup, setPatchItem, note, onAddItem, onPatchItem } = props;
  const [type, setType] = useState<string>("text");
  const [color, setColor] = useState<number>(-1);
  const [title, setTitle] = useState<string>("");
  const [text, setText] = useState<string>("");
  const [items, setItems] = useState<string[]>([]);
  const [image, setImage] = useState<string>("");
  const [activeTags, setTags] = useState<number[]>([]);
  const [reminder, setReminder] = useState<number>(0);
  const [attachType, setAttachType] = useState<string>(
    note && note.attachments && note.attachments[0] && note.attachments[0].type
      ? note.attachments[0].type
      : "link"
  );
  const [attachItems, setAttachItems] = useState<string[]>([]);

  const [checked, setChecked] = useState<boolean>(false);

  useEffect(() => {
    setAttachItems([]);
  }, [attachType]);

  useEffect(() => {
    if (note !== {}) {
      if (note.type) setType(note.type);
      if (note.title) setTitle(note.title);
      if (note.tags && note.tags.length > 0) setTags(note.tags);
      if (note.color !== undefined) setColor(note.color);
      if (note.items && note.items.length > 0) {
        let arr: string[] = [];
        note.items.forEach(item => {
          if (item.text !== undefined) {
            arr.push(item.text);
          }
        });
        setItems(arr);
      }
      if (note.text) setText(note.text);
      if (note.attachments) {
        setChecked(true);
        let arr: string[] = [];
        note.attachments.forEach(item => {
          if (item.url) arr.push(item.url);
        });
        setAttachItems(arr);
      }
      if (note.reminder) setReminder(note.reminder);
      if (note.url) setImage(note.url);
    }
  }, []);

  const inputRef = useRef<HTMLInputElement>(null);

  const onSubmitClick = (): void => {
    let newNote: INote = {};
    newNote.type = type;
    if (note.id !== undefined) newNote.id = note.id;
    if (title !== "") newNote.title = title;
    if (color !== -1) newNote.color = color;
    if (text !== "" && type === "text") newNote.text = text;
    if (items.length > 0 && type === "list") {
      newNote.items = items.map(item => {
        return { text: item, checked: false };
      });
    }
    if (image !== "" && type === "image") newNote.url = image;
    if (activeTags.length > 0) newNote.tags = activeTags;
    if (reminder !== 0) newNote.reminder = reminder;
    if (attachItems.length > 0) {
      newNote.attachments = attachItems.map(item => {
        return {
          type: attachType,
          url: item
        };
      });
    }
    setPatchItem({});
    if (newNote.id !== undefined) {
      onPatchItem(newNote);
    } else {
      onAddItem(newNote);
    }
  };

  return (
    <div className={styles.background}>
      <div className={styles.popup}>
        <button
          className={styles.exit}
          onClick={() => {
            setPatchItem({});
            setPopup(false);
          }}
        />
        <TypeList
          wrapper={styles.wrapper}
          type={type}
          name="type"
          onTypeChange={setType}
          items={[
            { text: "Список", value: "list" },
            { text: "Текст", value: "text" },
            { text: "Картинка", value: "image" }
          ]}
        />
        <ConstructorReminder reminder={reminder} setReminder={setReminder} />
        <label className={styles.title_wrapper}>
          <span className={styles.title}>Заголовок:</span>
          <input
            type="text"
            className={styles.title_input}
            ref={inputRef}
            value={title !== "" ? title : ""}
            onChange={() => {
              if (inputRef.current) {
                setTitle(inputRef.current.value);
              }
            }}
          />
        </label>

        <ConstructorColors
          wrapper={styles.wrapper}
          color={color}
          setColor={setColor}
        />
        {type === "text" && (
          <TextConstructor
            wrapper={styles.wrapper}
            text={text}
            setText={setText}
          />
        )}
        {type === "list" && (
          <ListConstructor
            wrapper={styles.wrapper}
            items={items}
            setItems={setItems}
          />
        )}
        {type === "image" && (
          <ImageConstructor
            image={image}
            setImage={setImage}
            wrapper={styles.wrapper}
          />
        )}
        <TagsList activeTags={activeTags} setTags={setTags} />

        <ConstructorAttachments
          type={attachType}
          name="attach"
          onTypeChange={setAttachType}
          items={attachItems}
          setAttachItems={setAttachItems}
          checked={checked}
          setChecked={setChecked}
        />
        <button
          className={styles.button}
          type="button"
          onClick={() => {
            setPopup(false);
            onSubmitClick();
          }}
        >
          Готово!
        </button>
      </div>
    </div>
  );
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<IState, null, AnyAction>
): IDispatchToProps => {
  return {
    onPatchItem: (note: INote) => {
      dispatch(patchNote(note));
    },
    onAddItem: (note: INote) => {
      dispatch(addNote(note));
    }
  };
};

export default connect(
  null,
  mapDispatchToProps
)(NoteConstructor);
