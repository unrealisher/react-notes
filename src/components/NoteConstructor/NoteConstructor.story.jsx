import React from "react";
import { Provider } from "react-redux";
import store from "./../../store/index";

import NoteConstructor from "./NoteConstructor";

import { storiesOf } from "@storybook/react";

import "./../../index.scss";

storiesOf("NoteConstructor", module).add("стандарт", () => (
  <Provider store={store}>
    <NoteConstructor />
  </Provider>
));
