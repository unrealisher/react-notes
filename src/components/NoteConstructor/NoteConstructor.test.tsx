import React from "react";
import { shallow } from "enzyme";
import { NoteConstructor } from "./NoteConstructor";

describe("NoteConstructor", () => {
  const onAddItem = (): void => {};
  const onPatchItem = (): void => {};
  const note = {
    id: 1,
    type: "text",
    title: "Завтра дедлайн",
    text: "Не забыть сверстать макет",
    tags: [1, 2],
    color: 2,
    size: "s",
    created: 1551714600000
  };
  const setPopup = (): void => {};
  const setPatchItem = (): void => {};

  it("correct render", () => {
    const component = shallow(
      <NoteConstructor
        onAddItem={onAddItem}
        onPatchItem={onPatchItem}
        note={note}
        setPopup={setPopup}
        setPatchItem={setPatchItem}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
