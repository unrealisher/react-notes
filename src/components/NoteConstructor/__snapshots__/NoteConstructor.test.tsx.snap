// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteConstructor correct render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NoteConstructor
    note={
      Object {
        "color": 2,
        "created": 1551714600000,
        "id": 1,
        "size": "s",
        "tags": Array [
          1,
          2,
        ],
        "text": "Не забыть сверстать макет",
        "title": "Завтра дедлайн",
        "type": "text",
      }
    }
    onAddItem={[Function]}
    onPatchItem={[Function]}
    setPatchItem={[Function]}
    setPopup={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="popup"
      >
        <button
          className="exit"
          onClick={[Function]}
        />
        <TypeList
          items={
            Array [
              Object {
                "text": "Список",
                "value": "list",
              },
              Object {
                "text": "Текст",
                "value": "text",
              },
              Object {
                "text": "Картинка",
                "value": "image",
              },
            ]
          }
          name="type"
          onTypeChange={[Function]}
          type="text"
          wrapper="wrapper"
        />
        <ConstructorReminder
          reminder={0}
          setReminder={[Function]}
        />
        <label
          className="title_wrapper"
        >
          <span
            className="title"
          >
            Заголовок:
          </span>
          <input
            className="title_input"
            onChange={[Function]}
            type="text"
            value=""
          />
        </label>
        <Connect(ConstructorColors)
          color={-1}
          setColor={[Function]}
          wrapper="wrapper"
        />
        <TextConstructor
          setText={[Function]}
          text=""
          wrapper="wrapper"
        />
        <Connect(TagsList)
          activeTags={Array []}
          setTags={[Function]}
        />
        <ConstructorAttachments
          checked={false}
          items={Array []}
          name="attach"
          onTypeChange={[Function]}
          setAttachItems={[Function]}
          setChecked={[Function]}
          type="link"
        />
        <button
          className="button"
          onClick={[Function]}
          type="button"
        >
          Готово!
        </button>
      </div>,
      "className": "background",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="exit"
            onClick={[Function]}
          />,
          <TypeList
            items={
              Array [
                Object {
                  "text": "Список",
                  "value": "list",
                },
                Object {
                  "text": "Текст",
                  "value": "text",
                },
                Object {
                  "text": "Картинка",
                  "value": "image",
                },
              ]
            }
            name="type"
            onTypeChange={[Function]}
            type="text"
            wrapper="wrapper"
          />,
          <ConstructorReminder
            reminder={0}
            setReminder={[Function]}
          />,
          <label
            className="title_wrapper"
          >
            <span
              className="title"
            >
              Заголовок:
            </span>
            <input
              className="title_input"
              onChange={[Function]}
              type="text"
              value=""
            />
          </label>,
          <Connect(ConstructorColors)
            color={-1}
            setColor={[Function]}
            wrapper="wrapper"
          />,
          <TextConstructor
            setText={[Function]}
            text=""
            wrapper="wrapper"
          />,
          false,
          false,
          <Connect(TagsList)
            activeTags={Array []}
            setTags={[Function]}
          />,
          <ConstructorAttachments
            checked={false}
            items={Array []}
            name="attach"
            onTypeChange={[Function]}
            setAttachItems={[Function]}
            setChecked={[Function]}
            type="link"
          />,
          <button
            className="button"
            onClick={[Function]}
            type="button"
          >
            Готово!
          </button>,
        ],
        "className": "popup",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "exit",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "items": Array [
              Object {
                "text": "Список",
                "value": "list",
              },
              Object {
                "text": "Текст",
                "value": "text",
              },
              Object {
                "text": "Картинка",
                "value": "image",
              },
            ],
            "name": "type",
            "onTypeChange": [Function],
            "type": "text",
            "wrapper": "wrapper",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "reminder": 0,
            "setReminder": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="title"
              >
                Заголовок:
              </span>,
              <input
                className="title_input"
                onChange={[Function]}
                type="text"
                value=""
              />,
            ],
            "className": "title_wrapper",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Заголовок:",
                "className": "title",
              },
              "ref": null,
              "rendered": "Заголовок:",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "title_input",
                "onChange": [Function],
                "type": "text",
                "value": "",
              },
              "ref": Object {
                "current": null,
              },
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "color": -1,
            "setColor": [Function],
            "wrapper": "wrapper",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "setText": [Function],
            "text": "",
            "wrapper": "wrapper",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "activeTags": Array [],
            "setTags": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "checked": false,
            "items": Array [],
            "name": "attach",
            "onTypeChange": [Function],
            "setAttachItems": [Function],
            "setChecked": [Function],
            "type": "link",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Готово!",
            "className": "button",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "Готово!",
          "type": "button",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="popup"
        >
          <button
            className="exit"
            onClick={[Function]}
          />
          <TypeList
            items={
              Array [
                Object {
                  "text": "Список",
                  "value": "list",
                },
                Object {
                  "text": "Текст",
                  "value": "text",
                },
                Object {
                  "text": "Картинка",
                  "value": "image",
                },
              ]
            }
            name="type"
            onTypeChange={[Function]}
            type="text"
            wrapper="wrapper"
          />
          <ConstructorReminder
            reminder={0}
            setReminder={[Function]}
          />
          <label
            className="title_wrapper"
          >
            <span
              className="title"
            >
              Заголовок:
            </span>
            <input
              className="title_input"
              onChange={[Function]}
              type="text"
              value=""
            />
          </label>
          <Connect(ConstructorColors)
            color={-1}
            setColor={[Function]}
            wrapper="wrapper"
          />
          <TextConstructor
            setText={[Function]}
            text=""
            wrapper="wrapper"
          />
          <Connect(TagsList)
            activeTags={Array []}
            setTags={[Function]}
          />
          <ConstructorAttachments
            checked={false}
            items={Array []}
            name="attach"
            onTypeChange={[Function]}
            setAttachItems={[Function]}
            setChecked={[Function]}
            type="link"
          />
          <button
            className="button"
            onClick={[Function]}
            type="button"
          >
            Готово!
          </button>
        </div>,
        "className": "background",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="exit"
              onClick={[Function]}
            />,
            <TypeList
              items={
                Array [
                  Object {
                    "text": "Список",
                    "value": "list",
                  },
                  Object {
                    "text": "Текст",
                    "value": "text",
                  },
                  Object {
                    "text": "Картинка",
                    "value": "image",
                  },
                ]
              }
              name="type"
              onTypeChange={[Function]}
              type="text"
              wrapper="wrapper"
            />,
            <ConstructorReminder
              reminder={0}
              setReminder={[Function]}
            />,
            <label
              className="title_wrapper"
            >
              <span
                className="title"
              >
                Заголовок:
              </span>
              <input
                className="title_input"
                onChange={[Function]}
                type="text"
                value=""
              />
            </label>,
            <Connect(ConstructorColors)
              color={-1}
              setColor={[Function]}
              wrapper="wrapper"
            />,
            <TextConstructor
              setText={[Function]}
              text=""
              wrapper="wrapper"
            />,
            false,
            false,
            <Connect(TagsList)
              activeTags={Array []}
              setTags={[Function]}
            />,
            <ConstructorAttachments
              checked={false}
              items={Array []}
              name="attach"
              onTypeChange={[Function]}
              setAttachItems={[Function]}
              setChecked={[Function]}
              type="link"
            />,
            <button
              className="button"
              onClick={[Function]}
              type="button"
            >
              Готово!
            </button>,
          ],
          "className": "popup",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "exit",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "items": Array [
                Object {
                  "text": "Список",
                  "value": "list",
                },
                Object {
                  "text": "Текст",
                  "value": "text",
                },
                Object {
                  "text": "Картинка",
                  "value": "image",
                },
              ],
              "name": "type",
              "onTypeChange": [Function],
              "type": "text",
              "wrapper": "wrapper",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "reminder": 0,
              "setReminder": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span
                  className="title"
                >
                  Заголовок:
                </span>,
                <input
                  className="title_input"
                  onChange={[Function]}
                  type="text"
                  value=""
                />,
              ],
              "className": "title_wrapper",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Заголовок:",
                  "className": "title",
                },
                "ref": null,
                "rendered": "Заголовок:",
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "title_input",
                  "onChange": [Function],
                  "type": "text",
                  "value": "",
                },
                "ref": Object {
                  "current": null,
                },
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "color": -1,
              "setColor": [Function],
              "wrapper": "wrapper",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "setText": [Function],
              "text": "",
              "wrapper": "wrapper",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "activeTags": Array [],
              "setTags": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "checked": false,
              "items": Array [],
              "name": "attach",
              "onTypeChange": [Function],
              "setAttachItems": [Function],
              "setChecked": [Function],
              "type": "link",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Готово!",
              "className": "button",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": "Готово!",
            "type": "button",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
